### ISSUE TITLE
Enum Nonmember Attribute Access Regression

### ISSUE THREAD

**Issue by max-muoto** [microsoft/pyright#7924]
There seems to be a regression with `enum.nonmember` after upgrading to 1.1.363:

```python
import enum

class MyEnum(enum.StrEnum):
    VAL_1 = "val_1"
    VAL_2 = "val_2"

    _items = enum.nonmember({VAL_1: 1, VAL_2: 2})

    def to_int(self) -> int:
        return self._items[self]

```

[Pyright Playground
](https://pyright-play.net/?pythonVersion=3.12&strict=true&code=JYWwDg9gTgLgBAUwHYFcQCh0GMA2BDAZwLgFkBPAUVRAApk0A6AZRiirQEoAudOPuAGoBBADIB9AIxwAvHABEANzw5Jc3v2HiATDPlKVWter5jgMBCGKz6IBkghIQFgEYIoNAN6bJXOBIA0gqJiWr5aAL4cmPxwACYIAGZwMBCmSDA0BAg4CRxwALQAfHDA6TwxMVAIMChQSHBZOQym5pYA2o0JALpAA)

Take this example, here Pyright isn't able to understand that `__getitem__` is an attribute on the non-member dictionary.

Version: 1.1.363

**Comment by max-muoto** [comment]
@erictraut, here's another example, where general type inference fails with dictionary comprehension, that used to work prior:

```python
import enum
from typing import reveal_type


class MyEnum(enum.StrEnum):
    _FROM_TYPE = enum.nonmember(
        {
            0: "INT",
            1: "STRING",
        }
    )
    _TO_TYPE = enum.nonmember({v: k for k, v in _FROM_TYPE.items()})


reveal_type(MyEnum._FROM_TYPE)
```

[Pyight Playground](https://pyright-play.net/?pythonVersion=3.12&strict=true&code=JYWwDg9gTgLgBAUwHYFcQCgBmUIjjATzGCQHM5RJY4oEA3BAQwBsB9QsBdbgY2cYDOAuAFkCAUVQgAFMjQA6AMowoktAEoAXOji64rAGIAlAPIjWAFQCaABXFwAvIinykEJCAQgARgijSdPSCAb0Cg8LgABk04ACIASQA5C1iAGjCIvQBGGNjFCyMkgHE0jKCAXwz1DMsTS1t7JzkQV3dPHz9pYLoYgGs4TGg4XtS4OgokfWMzert5YBgvAWl1curuWgYWdiIEaTE1FsNTc2s7aqA)

**Comment by erictraut** [comment]
The problem is that the typeshed class definitions of `member` and `nonmember` do not define a `__getattr__` method. If you'd like this to type check without modifications, you could try to convince the typeshed maintainers to add a `__getattr__` method.

```python
class nonmember(Generic[_EnumMemberT]):
    ...
    def __getattr__(self, name: str) -> Any: ...
```

This happened to work without a type error in versions of pyright prior to 1.1.363, but now pyright is more closely modeling the runtime behavior.

It's theoretically possible for a type checker to add special-case logic to handle a case like this, but I'm very reluctant to add such special-case logic, especially for such an uncommon construct. I'm therefore going to close this as "won't fix".

If you want this to work without type errors, you should change your code to `_FROM_TYPE.value.items()`.

**Comment by max-muoto** [comment]
> The problem is that the typeshed class definitions of `member` and `nonmember` do not define a `__getattr__` method. If you'd like this to type check without modifications, you could try to convince the typeshed maintainers to add a `__getattr__` method.
> 
> ```python
> class nonmember(Generic[_EnumMemberT]):
>     ...
>     def __getattr__(self, name: str) -> Any: ...
> ```
> 
> This happened to work without a type error in versions of pyright prior to 1.1.363, but now pyright is more closely modeling the runtime behavior.
> 
> It's theoretically possible for a type checker to add special-case logic to handle a case like this, but I'm very reluctant to add such special-case logic, especially for such an uncommon construct. I'm therefore going to close this as "won't fix".
> 
> If you want this to work without type errors, you should change your code to `_FROM_TYPE.value.items()`.

Sounds good, thanks for the information!

**Comment by max-muoto** [comment]
@erictraut, just to follow up, I face runtime errors when using `.value`, Pyright doesn't seem to complain though. Any suggestions here?:

**Comment by max-muoto** [comment]
It seems typeshed is inaccurate in providing `value`.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.364.
